File: src/leetcode/array/BestTimeToBuyAndSellStock.java
package leetcode.array;

// 다이나믹 프로그래밍 - Kadane's Algorithm을 이용한다.
// 거장 작은 값을 기억하고 있다가 현재 인덱스에서 팔았을 때의 profit을 비교하며 최대 값을 찾으면 된다.
public class BestTimeToBuyAndSellStock {
    class Solution {
        public int maxProfit(int[] prices) {
            int maxi = 0;
            int minimum = prices[0];
            for (int i=0; i< prices.length; i++ ){
                minimum = Math.min(minimum ,prices[i]);
                maxi = Math.max(maxi ,prices[i] - minimum );
            }
            return maxi;
        }
    }
}

File: src/leetcode/array/BestTimeToBuyAndSellStock.js
var maxProfit = function(prices) {
    let max = 0;
    let minNum = prices[0];
    for (let i = 0; i < prices.length; i++) {
        minNum = Math.min(minNum, prices[i]);
        max = Math.max(max, prices[i] - minNum);
    }
};

console.log(maxProfit([7, 1, 5, 3, 6, 4]));